### PR TITLE
 return value of method call can be object

### PR DESCRIPTION
**Pull request by dominikfeininger**
- still basic implementation
- to discuss: best way of remove circular
- todo: add tests to prove

for #224

### CODE REVIEW

**Comment by dominikfeininger** [comment]
@vobu pls review ;)

File: client-side-js/injectUI5.js [hunk 1]
```diff
@@ -236,6 +236,71 @@ async function clientSide_injectUI5(config, waitForUI5Timeout) {
...
+                                console.log(`removed ${typeof obj}`)
+                            }
+                            return obj
+                        }
+
+                        return detect(obj)
+                    }
+
+                    /**
+                     * JSON.stringify(circularReference, getCircularReplacer());
```
Comment (vobu) [inline client-side-js/injectUI5.js:288]: proper JSDoc pleae

File: client-side-js/executeControlMethod.js [hunk 2]
```diff
@@ -44,11 +46,23 @@ async function clientSide_executeControlMethod(webElement, methodName, args) {
                             if (window.wdi5.isPrimitive(result)) {
                                 // getter
                                 done(["success", result, "result"])
-                            } else {
+                            } else if (
+                                typeof result === "object" &&
+                                result !== null &&
+                                !(result instanceof sap.ui.core.Control)
```
Comment (vobu) [inline client-side-js/executeControlMethod.js:52]: could you src comment on the "why?" please?

File: client-side-js/injectUI5.js [hunk 3]
```diff
@@ -236,6 +236,71 @@ async function clientSide_injectUI5(config, waitForUI5Timeout) {
...
+                            if (typeof value === "object" && value !== null) {
+                                if (seen.has(value)) {
+                                    return
+                                }
+                                seen.add(value)
+                            }
+                            return value
+                        }
+                    }
+
```
Comment (vobu) [inline client-side-js/injectUI5.js:303]: is the next circular-thingie function still necessary/used or can its' usage be taken over by any of the above 3?

File: examples/ui5-js-app/webapp/test/e2e/basic.test.js [hunk 4]
```diff
@@ -61,4 +74,29 @@ describe("ui5 basic", () => {
...
+
+        titleSelector.forceSelect = true
+        const title = await browser.asControl(titleSelector).setTitle(newTitle).getTitle()
+        expect(title).toEqual(newTitle)
+
+        const titleFirstTime = await browser.asControl(titleSelector).setTitle(newTitle)
+        const titleAgain = await titleFirstTime.setTitle(newTitle)
+        const response = await titleAgain.getTitle()
+        expect(response).toEqual(newTitle)
+    })
```
Comment (vobu) [inline examples/ui5-js-app/webapp/test/e2e/basic.test.js:101]: could you please rephrase the "its" to better communicate that these refer to the control retrieval/value api?

File: src/lib/wdi5-control.ts [hunk 5]
```diff
@@ -319,6 +319,10 @@ export class WDI5Control {
                 // return $self after a called method of the wdi5 instance to allow method chaining
                 return this
             case "result":
+                const responseObj = result[3]
+                if (responseObj) {
```
Comment (vobu) [inline src/lib/wdi5-control.ts:323]: refactor to shorthand à la
const a = b || c

**Review by vobu — CHANGES_REQUESTED** [review]

**Review by vobu — APPROVED** [review]

**Pull request merged**